src/main/java/org/reactome/addlinks/referencecreators/EnsemblReferenceCreator.java

```diff
@@ -0,0 +1,115 @@
+     */
+    private Set<String> collectEnsemblIdentifiers(Map<String, Map<String, List<String>>> mappings, String sourceIdentifier, String biomartSpeciesName) {
+        String proteinToTranscriptsKey = biomartSpeciesName + "_proteinToTranscripts";
+        String proteinToGenesKey = biomartSpeciesName + "_proteinToGenes";
+        String uniprotToENSPKey = biomartSpeciesName + "_uniprotToProteins";
+        // Check that the 'uniprotToProteins' mapping exists and that the source instance's
+        // 'identifier' attribute maps to ensembl protein identifiers.
+        Set<String> ensemblIds = new HashSet<>();
+        if (identifierHasMapping(mappings, uniprotToENSPKey, sourceIdentifier)) {
+            // Retrieve all ensembl protein identifiers that map to the source instance identifier and store in 'ensIds'.
+            List<String> enspIds = mappings.get(uniprotToENSPKey).get(sourceIdentifier);
+            for (String enspId : enspIds) {
+                // Only store Ensembl protein identifiers.
+                if (enspId.startsWith("ENS")) {
+                    ensemblIds.add(enspId);
+                }
+                // Retrieve any Ensembl transcript identifiers that map to the Ensembl protein identifier and store in 'ensIds'.
+                if (identifierHasMapping(mappings, proteinToTranscriptsKey, enspId)) {
+                    ensemblIds.addAll(collectEnsemblIdentifiers(mappings.get(proteinToTranscriptsKey).get(enspId)));
+                }
+                // Retrieve any Ensembl gene identifiers that map to the Ensembl protein identifier and store in 'ensIds'.
+                if (identifierHasMapping(mappings, proteinToGenesKey, enspId)) {
+                    ensemblIds.addAll(collectEnsemblIdentifiers(mappings.get(proteinToGenesKey).get(enspId)));
+                }
+
+            }
+        }
+        return ensemblIds;
+    }
+
+    // Filter non-Ensembl identifiers from supplied List.
+    private Collection<String> collectEnsemblIdentifiers(List<String> ensIds) {
+        Set<String> filteredEnsemblIds = new HashSet<>();
+        for (String ensId : ensIds) {
+            // Only store 'Ensembl' identifiers.
+            if (ensId.startsWith("ENS")) {
+                filteredEnsemblIds.add(ensId);
+            }
+        }
+        return filteredEnsemblIds;
+    }
+
+
+    // Check that the supplied identifier has a mapping in the 'mappings' structure.
+    private boolean identifierHasMapping(Map<String, Map<String, List<String>>> mappings, String outerKey, String identifier) {
+        return mappings.containsKey(outerKey) && mappings.get(outerKey).containsKey(identifier);
+    }
+
+    // Update species name attribute to Ensembl Biomart format (eg: Homo sapiens --> hsapiens).
+    private String getBiomartSpeciesName(String speciesName) {
+        return speciesName.substring(0,1).toLowerCase() + speciesName.split(" ")[1];
+    }
```

Comment:
I've seen this logic in a few places in the codebase.  Maybe moving this to a utilities class that could be used wherever this logic is needed would be good?

Reply:
Done